### ISSUE TITLE
Discussion: how to ship vanilla data (recipes, loot tables) in Feather

### ISSUE THREAD

**Issue by caelunshun** [feather-rs/feather#267]
Loot tables and recipes are loaded from the vanilla data pack, which is extracted from the vanilla server JAR through `feather-data`. Currently, they are embedded in the server executable as JSON. This raises a number of legal issues, since we're directly packaging and redistributing Mojang assets. I see three potential solutions:

* Transcode the data from JSON to another format, e.g. CBOR. The assets will be distributed in a different format, making the legal violation slightly less blatant. However, I doubt this is legal either.
* Download the vanilla data pack at server startup. This will completely avoid packaging Mojang assets, but it's not very convenient for users to wait for the download to complete. Feather won't feel as "lightweight" when its first startup potentially takes a minute or two.
* Manually write the recipes and loot tables in our own format without consulting the vanilla assets. This will be extremely tedious, but it is the most ideal option in terms of results.

If anyone else has another idea, please share it here—none of the options above seem all that great to me :)

**Comment by Schuwi** [comment]
I mean, ideally it will always be the same data in the end, so I guess for the legal aspect the thing that matters is where it comes from?
So at least for recipes community sources like https://minecraft.gamepedia.com/Crafting#Complete_recipe_list might be an idea but I doubt we'd find any community sourced loot tables that aren't just extracted from the Vanilla JAR.

Sooo... I had another idea, which is most certainly overkill and not really practical but I guess it's worth at least mentioning...
If one were to use a bot to connect to a vanilla server, run automated tests to source the info needed (again easy for recipes as they are just sent to the client I believe? but probably very hard for the loot table) and generate a handy file...

Pros:
- This should at least be perfectly legal (unless writing bots for Minecraft is generally forbidden or something, which could be the case as I've actually never read the EULA to be honest lol)
- No inconvenience for the user yay!!
- No need for error-prone laborious manual updating of recipes on updates (or in general)

Cons:
- Well... I mean no sane person would do this right?
- It's a *lot* of effort. though there are probably at least existing implementations for general purpose Minecraft bots that could be appropriated
- Instead of having to update a list of recipes you now have to update a whole client, yeah! (though it's probably mostly protocol which has to be updated anyway)

There are probably crazier things that have been done buuuut this would definitely require a good amount of desperation :D

**Comment by Schuwi** [comment]
I just found out we get some of our stuff (like block states) by letting the vanilla server jar generate the (json?) files for us (https://wiki.vg/Data_Generators).
Shouldn't that be alright to redistribute? As I would compare this to world saves (also generated willingly by official Minecraft software) which are freely distributed online and it's not like extracting files from the server jar.

**Comment by Schuwi** [comment]
Hm, Mojang actually doesn't explicitly mention anything about files generated by the game (like world saves) in any of their legal stuff.

https://account.mojang.com/terms#brand
This is about game assets but this seems to only apply to stuff actually "in the game" like in the jar file.

https://account.mojang.com/documents/minecraft_eula
From the EULA I'm not sure whether feather would fall under their definition of a mod

    By "Mods", we mean something original that you or someone else created that doesn't contain **a substantial part** of our copyrightable code or content.
(emphasis mine)

So yeah, these generated reports are probably not on the same level as world saves, as they don't include any of your original content (buildings and stuff) like a world save would. So the copyright for those files would most certainly still fall to Mojang. Though as they are also not part **of** they game, they shouldn't fall under the strict rules of the Brand and Assets legal document.
So in case feather is classified as a "Mod" per their definition I would argue that using these reports, feather could be classified as containing a non-substantial part of Mojang's copyrightable content.

**Comment by Schuwi** [comment]
Okay, I haven't really thought about the feasibility of this yet but here is my idea:
- We don't extract files from the JAR
- Instead we only use the data generators during build with the downloaded JAR, like it's already being done now with some of the generators
- Also community datasets like minecraft-data

And transform all this to build-time rust code. No run-time parsing of JSON files or something. Basically building a database of all recipes, block properties, entity drops, item properties, etc... maybe using stuff like phf to build const hashsets and make it statically queryable. What I mean with that is I want to be able to write something akin to a SQL query to e.g. give me all blocks that are not solid but conduct redstone or something which will be statically resolved at compile time to a const array of the results and if possible even a match arm via macro. So that it's not only possible to check whether a specific BlockKind/BlockId has a specific property but also to get all with a specific property.
It should also be trivial to add new properties to synthesize from the input data, so that we can further abstract from the original Minecraft data and implement it like we need it. Maybe even rewrite block states to abstract them further and somehow implement them with that system? The details are fuzzy here, I haven't really thought about that yet.
Basically a complete rewrite of many of the core crates.

Well, that would be my dream anyway. Having all "hard-coded" magic values in a central place and being able to run abstract queries at run-time on that data set. Next I'll look into the feasibility and have my dreams brutally crushed :)

Ah yes and only using the data generators indirectly (for the build process) should be enough to avoid any legal problems I think, in accordance with my arguments in previous comments.

**Comment by Schuwi** [comment]
Okay, maybe I've gone a bit overboard here. What I want to advocate:
- No extraction of files from the JAR
- Conversion of all data to pure Rust during build

I think we should be good from a legal standpoint with that per my previous arguments.

**Comment by Inky-developer** [comment]
I think that would mean that some data would have to be created manually. For example (as far as i know) the data generators create no data about what blocks can conduct strong redstone power, so a list of these blocks would have to be manually created and updated for each new version.

**Comment by Schuwi** [comment]
Yes, that's a good point. Though we would have to do that either way because afaik many of these properties are just hard-coded in Minecraft and are probably even harder to extract from the source code than to enter manually.

**Comment by Inky-developer** [comment]
I guess something like this could be used then: https://github.com/PrismarineJS/minecraft-data

**Comment by Schuwi** [comment]
Yeah, we are already using `minecraft-data`(see `feather-data`) but afaik that is also missing a lot of those kinds of hard-coded properties.

**Comment by Defman** [comment]
We already use PrismarineJS for some of its data. The Minecraft jar contains achievements, recipes and etc in JSON, this information is not exposed through the Minecraft data generator; the data generator is mostly used for block state ids.

**Comment by Inky-developer** [comment]
If I run the data generator with the --all flag it generates all advancements, loot tables, etc. in the data/minecraft folder

**Comment by Defman** [comment]
MB, still languages are missing, which are going to be used in the text api.

**Comment by Schuwi** [comment]
Hmm yeah the loot tables might be a problem though. They can't be generated by the data generators and are also not in `minecraft-data`.

**Comment by pielover928** [comment]
You guys will want to look into [Burger](https://github.com/pielover928/Burger), which takes in a .jar and extracts recipes/blockstates/etc. and puts it into a JSON. It doesn't have a "topping" for loot tables, but it's easily extendable.

**Comment by Schuwi** [comment]
Well, extracting the data is not really the problem here, it is already implemented right now. The problem is mostly the legal aspect, which wouldn't change just by using external dependencies for data extraction.

**Comment by pielover928** [comment]
As long as you require users to agree to the EULA when they create a server (the same way a vanilla server does,) you're essentially only acting as a modding tool, which is very much allowed under the EULA: https://account.mojang.com/documents/minecraft_eula

While they have final say in deciding whether or not something is considered a mod, they essentially ask that you try to make sure that users are accessing the game in an official way (i.e. agreed to the EULA) and that you don't try to cause damage to the game's servers or community.

**Comment by Schuwi** [comment]
We could implement asset loading in a generic way such that it can happen at either runtime or buildtime - downloading the assets directly from Mojang in either case. Use a feature flag to control this which is set to buildtime asset loading by default so anyone building feather themselves with `cargo install feather` will get the no-warmup build while we only provide binaries built with runtime asset loading to avoid any legal issues.

I'm probably missing something because this solution sounds too perfect, right?
